File: src/core/parse/value.ts
import { prefixMultiplier, type Unit } from "../units.js";

export type ParsedValue =
  | {
      kind: "quantity";
      /** In base units (ohms / farads / henries), whichever the class supplies. */
      magnitude: number;
      /** Set only when the string itself said so (`100nF`, `10Ω`), not inferred. */
      unit: Unit | undefined;
      /** True when the string carried neither unit nor SI prefix: `100`, `0`, `4`. */
      bare: boolean;
      raw: string;
    }
  | {
      /** Not a quantity: an MPN, a placeholder, a name. `RP2040`, `LED`, `575-4`. */
      kind: "opaque";
      raw: string;
    };

/** Unicode and spelling variants that mean the same thing. */
function normalize(raw: string): string {
  return raw
    .trim()
    .replace(/µ|μ/g, "u") // MICRO SIGN, GREEK SMALL LETTER MU
    .replace(/Ω|Ω/g, "ohm") // OHM SIGN, GREEK CAPITAL OMEGA
    .replace(/\s+/g, "");
}

/** Strips an explicit trailing unit, returning what it was. */
function takeUnit(s: string): { rest: string; unit: Unit | undefined } {
  const tail = [
    { re: /(?:ohms?)$/i, unit: "ohm" as const },
    { re: /(?:farads?)$/i, unit: "farad" as const },
    { re: /(?:henrys?|henries)$/i, unit: "henry" as const },
    { re: /F$/, unit: "farad" as const },
    { re: /H$/, unit: "henry" as const },
  ];
  for (const { re, unit } of tail) {
    const rest = s.replace(re, "");
    // Only a unit if something numeric survives; guards `F` in a part number.
    if (rest !== s && /\d/.test(rest)) return { rest, unit };
  }
  return { rest: s, unit: undefined };
}

const rkm = /^(\d+)([a-zA-Zµ])(\d+)$/; //  5k6, 3n3, 0R1, 4u7
const plain = /^(\d+(?:\.\d+)?)([a-zA-Zµ]?)$/; //  10k, 4.7n, 100, 0.1

/**
 * Parses a KiCad `Value` field into a magnitude, or reports it as opaque.
 *
 * The unit is usually absent (`10k`), so the caller supplies it from the
 * component class; `unit` here is only what the text itself asserted.
 */
export function parseValue(raw: string): ParsedValue {
  const trimmed = raw.trim();
  if (trimmed === "") return { kind: "opaque", raw };

  const { rest, unit } = takeUnit(normalize(trimmed));

  const rkmMatch = rkm.exec(rest);
  if (rkmMatch) {
    const [, whole, letter, frac] = rkmMatch as unknown as [string, string, string, string];
    const multiplier = prefixMultiplier(letter);
    if (multiplier !== undefined) {
      const magnitude = Number(`${whole}.${frac}`) * multiplier;
      if (Number.isFinite(magnitude))
        return { kind: "quantity", magnitude, unit, bare: false, raw };
    }
    return { kind: "opaque", raw };
  }

  const plainMatch = plain.exec(rest);
  if (plainMatch) {
    const [, digits, letter] = plainMatch as unknown as [string, string, string];
    const multiplier = letter === "" ? 1 : prefixMultiplier(letter);
    if (multiplier !== undefined) {
      const magnitude = Number(digits) * multiplier;
      if (Number.isFinite(magnitude)) {
        return {
          kind: "quantity",
          magnitude,
          unit,
          bare: letter === "" && unit === undefined,
          raw,
        };
      }
    }
    return { kind: "opaque", raw };
  }

  return { kind: "opaque", raw };
}

File: kinv/core/parse/value.py
"""Reading a KiCad ``Value`` field as a magnitude, or recognising that it is not one."""

from __future__ import annotations

import math
import re
from typing import Any

from kinv.core.units import prefix_multiplier

ParsedValue = dict[str, Any]
"""``{kind: "quantity", magnitude, unit, bare, raw}`` or ``{kind: "opaque", raw}``."""


def _normalize(raw: str) -> str:
    """Unicode and spelling variants that mean the same thing."""
    text = raw.strip()
    text = re.sub("[µμ]", "u", text)  # MICRO SIGN, GREEK SMALL LETTER MU
    text = re.sub("[ΩΩ]", "ohm", text)  # OHM SIGN, GREEK CAPITAL OMEGA
    return re.sub(r"\s+", "", text)


_TAILS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"(?:ohms?)\Z", re.IGNORECASE), "ohm"),
    (re.compile(r"(?:farads?)\Z", re.IGNORECASE), "farad"),
    (re.compile(r"(?:henrys?|henries)\Z", re.IGNORECASE), "henry"),
    (re.compile(r"F\Z"), "farad"),
    (re.compile(r"H\Z"), "henry"),
]


def _take_unit(text: str) -> tuple[str, str | None]:
    """Strips an explicit trailing unit, returning what it was."""
    for pattern, unit in _TAILS:
        rest = pattern.sub("", text, count=1)
        # Only a unit if something numeric survives; guards `F` in a part number.
        if rest != text and re.search("[0-9]", rest):
            return rest, unit
    return text, None


_RKM = re.compile("^([0-9]+)([a-zA-Zµ])([0-9]+)\\Z")  # 5k6, 3n3, 0R1, 4u7
_PLAIN = re.compile("^([0-9]+(?:\\.[0-9]+)?)([a-zA-Zµ]?)\\Z")  # 10k, 4.7n, 100, 0.1


def parse_value(raw: str) -> ParsedValue:
    """Parses a KiCad ``Value`` field into a magnitude, or reports it as opaque.

    The unit is usually absent (``10k``), so the caller supplies it from the
    component class; ``unit`` here is only what the text itself asserted.
    """
    trimmed = raw.strip()
    if trimmed == "":
        return {"kind": "opaque", "raw": raw}

    rest, unit = _take_unit(_normalize(trimmed))

    rkm = _RKM.match(rest)
    if rkm:
        whole, letter, frac = rkm.groups()
        multiplier = prefix_multiplier(letter)
        if multiplier is not None:
            magnitude = float(f"{whole}.{frac}") * multiplier
            if math.isfinite(magnitude):
                return {"kind": "quantity", "magnitude": magnitude, "unit": unit, "bare": False, "raw": raw}
        return {"kind": "opaque", "raw": raw}

    plain = _PLAIN.match(rest)
    if plain:
        digits, letter = plain.groups()
        multiplier = 1.0 if letter == "" else prefix_multiplier(letter)
        if multiplier is not None:
            magnitude = float(digits) * multiplier
            if math.isfinite(magnitude):
                return {
                    "kind": "quantity",
                    "magnitude": magnitude,
                    "unit": unit,
                    "bare": letter == "" and unit is None,
                    "raw": raw,
                }
        return {"kind": "opaque", "raw": raw}

    return {"kind": "opaque", "raw": raw}
